### DISCUSSION TITLE
Best bezier implementation

### DISCUSSION THREAD

**Discussion by arkypita** [arkypita/LaserGRBL#1596] [General]
Hi @cdibbs and @darrenstarr and everyone else that want to contribute

In v4.5.1 I merge @darrenstarr code commit #1543 that implement a smart Bezier curve approximation. I found it fine because it looks giving better results in term of smoothness  of curve than the previous code. 

```
/// <summary>
/// For a series of Bezier curves represented as a series of control points, interpolate a series of points for representing the curve as lines
/// </summary>
/// 
/// Cubic bezier curves are typically represented as a list of 4 control points. Multiple continuous curves are concenated so that
/// the last point on one curve is the first point of the next.
/// 
/// This function takes a contiguous series of cubic bezier curves and returns an approximated representation of the curve as contiguous lines.
/// 
/// The error in this function is a measure of curviness. As such, segments are produced once the segments are declared flat enough. The smaller the
/// error, the flatter the curve must be before it's converted into a line segment.
/// 
/// <param name="points">The list of contiguous bezier segments</param>
/// <param name="error">The flatness error. 0.01 being the default, but 0.001 producing beautiful results.</param>
/// <returns>An enumerable list of points representing the line segments</returns>
public static IEnumerable<Point> FlattenTo(IList<Point> points, double error=0.01)
{
    var result = new List<Point> { points.First() };
    for (var i = 0; i + 3 <= points.Count; i += 3)
        result.AddRange(FlattenSegmentTo(points.Skip(i).Take(4).ToList(), error).Skip(1));

    return result;
}
```

I felt the need to decrease the value of the parameter "error" from the default value 0.01 to 0.5 because otherwise the algorithm produced an excessive number of points (and therefore of gcode segments).

@cdibbs fix an overflow error and now the code is fine.

However I am not satisfied with the result of this code, and I'm thinking of going back to the previous algorithm..
Using error 0.5 seems to give too crude a result in some cases, see #1590 issue
Conversely, using a lower error value unnecessarily increases the number of points in other cases.

I hypothesize that there could be a possibility to make the algorithm adaptive to the type of curve, applying different error values on different curves, but I don't know how to do it (I have no expertise in this area). Do you have any suggestions?

**Comment by arkypita** [comment]
Ps. I push a commit with the old code back, so it is possible to test and compare results.

**Comment by arkypita** [comment]
Well, I did some more test.
The "smart bezier" is already an adaptive  algorithm (while the old one was not). The only issue is that 0.5 is too much as error.
The issue with #1590 is because of this.

If I use 0.1 as error all is fine, except the number of point increase, but it is the only way to have a smoother curve.

**Reply by cdibbs** [reply]
Yeah, it makes sense that it would already be adaptive. If I'm understanding correctly, the `error` return value represents a distance in millimeters, so 0.1 would be a good value. It might be worth exposing that in the settings in case users want to experiment with this.

A pie-in-the-sky wish that just occurred to me: a bezier debugger. It would allow you to see areas of linear density (# segments per unit distance) as colors ranging from green to orange to red (very dense). Not sure that's appropriate for LaserGRBL since you would be tempted to edit in place, and editing seems out-of-scope. Maybe Inkscape? Ever heard of anything like that?

Anyway, thanks for your work on this!

**Comment by cdibbs** [comment]
@arkypita I came back to this later in the day to add some tests around the public `FlattenTo(...)` method. I wanted to satisfy my curiosity about some other edge cases. Unfortunately, that led me to discover that my overflow fix was more of a band-aid.

For examples which still fail, you only need to look at where a, b, and c are calculated at the start of the `CalculateFlatnessError` method. Any combination of values that results in them being zero will cause the overflow in the original code. Now, with my band-aid, there is no overflow, but it returns 0 instead of NaN which only works sometimes. I've attached a zipped svg showing the new issue: [bezier-simple.zip](https://github.com/arkypita/LaserGRBL/files/7274946/bezier-simple.zip). It contains only one bezier segment defined by a 100 x 100 square, corners enumerated clockwise. Inkscape renders it correctly as an arch. LaserGRBL with my band-aid renders it as flat lines (at least it doesn't crash :-).

On the bright side, I think the `CalculateFlatnessError` method itself may be unnecessary. I refactored `FlattenSegmentTo(...)` in order to always run `SplitCurveAt(...)` experimentally and examine whether the split contributed anything (by checking the area of the triangle formed by the current A-D line segment and the two, post-split segments). It seems to work in all the cases I've thrown at it, both manually and in unit tests, but it may need some more vetting. My latest PR has been updated with it. Thanks!

**Reply by cdibbs** [reply]
...but it's not optimal due to a misunderstanding on my part. Anyway, actual job begins now, so I'll look at it more later. :-)

**Reply by cdibbs** [reply]
@arkypita I got a chance to work on this a bit, today. All crashes and optimization problems seem to be fixed.

The biggest change is the removal of `CalculateFlatnessError` in BezierTools.cs, effectively replacing it with `CalculateTriangleArea`. This avoids the old edge cases and bugs since the new method is stateless and uses the square root of a sum of squares 
(which is always positive). The recursive case remains the same.

The effect of this seems to be a big reduction in generated GCode for a lot of the SVGs I tried. In fact, in most cases, it resulted in less code than the old approximation algorithm while still looking better. In some contrived cases, it produces more code but still looks better:

[bezier-simple.zip](https://github.com/arkypita/LaserGRBL/files/7317995/bezier-simple.zip)

Still, it might be worth making some of this into config options. Anyway, let me know what you think. Thanks!

**Comment by cdibbs** [comment]
@arkypita I saw you pulled in my PR. Do you think it would make sense to add an SVG Import settings tab next to the Raster one? If so, are you planning to work on that, or should I give it a go? Thanks!

**Reply by arkypita** [reply]
Yes I merged it. Is it stable, isn't it?
No I don't think to put a switch in settings. If it work and it gives better results I believe noone need to turn back to old mode.
What do you think?

**Reply by cdibbs** [reply]
Yeah, it should be stable, but I was thinking it would be good to avoid breaking anyone's existing workflows. Imagine someone out there is producing widgets that now have a certain look coupled to the legacy algorithm, and they wish to continue with that look. It doesn't cost the project much to keep the old code there along with the caveat that it's legacy and likely won't be getting as much developer attention going forward.

I'm also not confident that the 0.2 error value will work best in all scenarios, especially with very elaborate SVGs, or SVGs produced by software other than Inkscape (which is really the only tool I use for the moment). So, I thought it would be good to expose that, too.
